File: archive/pulumi-ts/Pr6618/lib/components/vpc-component.ts
import * as pulumi from '@pulumi/pulumi';
import * as aws from '@pulumi/aws';

export interface VpcComponentArgs {
  environmentSuffix: string;
  vpcCidr: string;
  availabilityZones: string[];
  publicSubnetCidrs: string[];
  privateSubnetCidrs: string[];
  enableDnsHostnames?: boolean;
  enableDnsSupport?: boolean;
}

export class VpcComponent extends pulumi.ComponentResource {
  public readonly vpc: aws.ec2.Vpc;
  public readonly publicSubnets: aws.ec2.Subnet[];
  public readonly privateSubnets: aws.ec2.Subnet[];
  public readonly internetGateway: aws.ec2.InternetGateway;
  public readonly publicRouteTable: aws.ec2.RouteTable;
  public readonly privateRouteTables: aws.ec2.RouteTable[];
  public readonly vpcId: pulumi.Output<string>;
  public readonly publicSubnetIds: pulumi.Output<string>[];
  public readonly privateSubnetIds: pulumi.Output<string>[];

  constructor(
    name: string,
    args: VpcComponentArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('custom:network:VpcComponent', name, {}, opts);

    // Create VPC
    this.vpc = new aws.ec2.Vpc(
      `vpc-${args.environmentSuffix}`,
      {
        cidrBlock: args.vpcCidr,
        enableDnsHostnames: args.enableDnsHostnames ?? true,
        enableDnsSupport: args.enableDnsSupport ?? true,
        tags: {
          Name: `vpc-${args.environmentSuffix}`,
          Environment: args.environmentSuffix,
        },
      },
      { parent: this }
    );

    this.vpcId = this.vpc.id;

    // Create Internet Gateway
    this.internetGateway = new aws.ec2.InternetGateway(
      `igw-${args.environmentSuffix}`,
      {
        vpcId: this.vpc.id,
        tags: {
          Name: `igw-${args.environmentSuffix}`,
          Environment: args.environmentSuffix,
        },
      },
      { parent: this }
    );

    // Create Public Subnets
    this.publicSubnets = args.publicSubnetCidrs.map((cidr, index) => {
      return new aws.ec2.Subnet(
        `public-subnet-${index}-${args.environmentSuffix}`,
        {
          vpcId: this.vpc.id,
          cidrBlock: cidr,
          availabilityZone: args.availabilityZones[index],
          mapPublicIpOnLaunch: true,
          tags: {
            Name: `public-subnet-${index}-${args.environmentSuffix}`,
            Environment: args.environmentSuffix,
            Type: 'public',
          },
        },
        { parent: this }
      );
    });

    this.publicSubnetIds = this.publicSubnets.map(subnet => subnet.id);

    // Create Private Subnets
    this.privateSubnets = args.privateSubnetCidrs.map((cidr, index) => {
      return new aws.ec2.Subnet(
        `private-subnet-${index}-${args.environmentSuffix}`,
        {
          vpcId: this.vpc.id,
          cidrBlock: cidr,
          availabilityZone: args.availabilityZones[index],
          mapPublicIpOnLaunch: false,
          tags: {
            Name: `private-subnet-${index}-${args.environmentSuffix}`,
            Environment: args.environmentSuffix,
            Type: 'private',
          },
        },
        { parent: this }
      );
    });

    this.privateSubnetIds = this.privateSubnets.map(subnet => subnet.id);

    // Create Public Route Table
    this.publicRouteTable = new aws.ec2.RouteTable(
      `public-rt-${args.environmentSuffix}`,
      {
        vpcId: this.vpc.id,
        tags: {
          Name: `public-rt-${args.environmentSuffix}`,
          Environment: args.environmentSuffix,
        },
      },
      { parent: this }
    );

    // Create route to Internet Gateway
    new aws.ec2.Route(
      `public-route-${args.environmentSuffix}`,
      {
        routeTableId: this.publicRouteTable.id,
        destinationCidrBlock: '0.0.0.0/0',
        gatewayId: this.internetGateway.id,
      },
      { parent: this }
    );

    // Associate Public Subnets with Public Route Table
    this.publicSubnets.forEach((subnet, index) => {
      new aws.ec2.RouteTableAssociation(
        `public-rta-${index}-${args.environmentSuffix}`,
        {
          subnetId: subnet.id,
          routeTableId: this.publicRouteTable.id,
        },
        { parent: this }
      );
    });

    // Create Private Route Tables (one per AZ)
    this.privateRouteTables = args.availabilityZones.map((az, index) => {
      return new aws.ec2.RouteTable(
        `private-rt-${index}-${args.environmentSuffix}`,
        {
          vpcId: this.vpc.id,
          tags: {
            Name: `private-rt-${index}-${args.environmentSuffix}`,
            Environment: args.environmentSuffix,
            AvailabilityZone: az,
          },
        },
        { parent: this }
      );
    });

    // Associate Private Subnets with Private Route Tables
    this.privateSubnets.forEach((subnet, index) => {
      new aws.ec2.RouteTableAssociation(
        `private-rta-${index}-${args.environmentSuffix}`,
        {
          subnetId: subnet.id,
          routeTableId: this.privateRouteTables[index].id,
        },
        { parent: this }
      );
    });

    this.registerOutputs({
      vpcId: this.vpcId,
      publicSubnetIds: this.publicSubnetIds,
      privateSubnetIds: this.privateSubnetIds,
    });
  }
}

File: archive/pulumi-py/Pr7062/lib/vpc_component.py
"""
VPC Component for Payment Processing Infrastructure.
Creates isolated VPC with public and private subnets.
"""

from typing import Optional, Dict
import pulumi
import pulumi_aws as aws
from pulumi import ResourceOptions, Output


class VpcComponentArgs:
    """Arguments for VPC Component."""

    def __init__(
        self,
        environment_suffix: str,
        cidr_block: str = "10.0.0.0/16",
        availability_zones: list = None,
        tags: Optional[Dict[str, str]] = None
    ):
        self.environment_suffix = environment_suffix
        self.cidr_block = cidr_block
        self.availability_zones = availability_zones or ["us-east-1a", "us-east-1b"]
        self.tags = tags or {}


class VpcComponent(pulumi.ComponentResource):
    """
    Reusable VPC component with isolated network configuration.
    Creates VPC with 2 public and 2 private subnets, IGW, and NAT Gateway.
    """

    def __init__(
        self,
        name: str,
        args: VpcComponentArgs,
        opts: Optional[ResourceOptions] = None
    ):
        super().__init__('payment:network:VpcComponent', name, None, opts)

        child_opts = ResourceOptions(parent=self)

        # Create VPC
        self.vpc = aws.ec2.Vpc(
            f"payment-vpc-{args.environment_suffix}",
            cidr_block=args.cidr_block,
            enable_dns_hostnames=True,
            enable_dns_support=True,
            tags={
                **args.tags,
                'Name': f"payment-vpc-{args.environment_suffix}",
            },
            opts=child_opts
        )

        # Create Internet Gateway
        self.igw = aws.ec2.InternetGateway(
            f"payment-igw-{args.environment_suffix}",
            vpc_id=self.vpc.id,
            tags={
                **args.tags,
                'Name': f"payment-igw-{args.environment_suffix}",
            },
            opts=child_opts
        )

        # Create public subnets
        self.public_subnets = []
        for i, az in enumerate(args.availability_zones[:2]):
            subnet = aws.ec2.Subnet(
                f"payment-public-subnet-{i+1}-{args.environment_suffix}",
                vpc_id=self.vpc.id,
                cidr_block=f"10.0.{i}.0/24",
                availability_zone=az,
                map_public_ip_on_launch=True,
                tags={
                    **args.tags,
                    'Name': f"payment-public-subnet-{i+1}-{args.environment_suffix}",
                    'Type': 'public',
                },
                opts=child_opts
            )
            self.public_subnets.append(subnet)

        # Create private subnets
        self.private_subnets = []
        for i, az in enumerate(args.availability_zones[:2]):
            subnet = aws.ec2.Subnet(
                f"payment-private-subnet-{i+1}-{args.environment_suffix}",
                vpc_id=self.vpc.id,
                cidr_block=f"10.0.{i+10}.0/24",
                availability_zone=az,
                map_public_ip_on_launch=False,
                tags={
                    **args.tags,
                    'Name': f"payment-private-subnet-{i+1}-{args.environment_suffix}",
                    'Type': 'private',
                },
                opts=child_opts
            )
            self.private_subnets.append(subnet)

        # Create Elastic IP for NAT Gateway
        self.eip = aws.ec2.Eip(
            f"payment-nat-eip-{args.environment_suffix}",
            domain="vpc",
            tags={
                **args.tags,
                'Name': f"payment-nat-eip-{args.environment_suffix}",
            },
            opts=child_opts
        )

        # Create NAT Gateway in first public subnet
        self.nat_gateway = aws.ec2.NatGateway(
            f"payment-nat-{args.environment_suffix}",
            subnet_id=self.public_subnets[0].id,
            allocation_id=self.eip.id,
            tags={
                **args.tags,
                'Name': f"payment-nat-{args.environment_suffix}",
            },
            opts=child_opts
        )

        # Create public route table
        self.public_route_table = aws.ec2.RouteTable(
            f"payment-public-rt-{args.environment_suffix}",
            vpc_id=self.vpc.id,
            tags={
                **args.tags,
                'Name': f"payment-public-rt-{args.environment_suffix}",
            },
            opts=child_opts
        )

        # Create route to Internet Gateway
        aws.ec2.Route(
            f"payment-public-route-{args.environment_suffix}",
            route_table_id=self.public_route_table.id,
            destination_cidr_block="0.0.0.0/0",
            gateway_id=self.igw.id,
            opts=child_opts
        )

        # Associate public subnets with public route table
        for i, subnet in enumerate(self.public_subnets):
            aws.ec2.RouteTableAssociation(
                f"payment-public-rta-{i+1}-{args.environment_suffix}",
                subnet_id=subnet.id,
                route_table_id=self.public_route_table.id,
                opts=child_opts
            )

        # Create private route table
        self.private_route_table = aws.ec2.RouteTable(
            f"payment-private-rt-{args.environment_suffix}",
            vpc_id=self.vpc.id,
            tags={
                **args.tags,
                'Name': f"payment-private-rt-{args.environment_suffix}",
            },
            opts=child_opts
        )

        # Create route to NAT Gateway
        aws.ec2.Route(
            f"payment-private-route-{args.environment_suffix}",
            route_table_id=self.private_route_table.id,
            destination_cidr_block="0.0.0.0/0",
            nat_gateway_id=self.nat_gateway.id,
            opts=child_opts
        )

        # Associate private subnets with private route table
        for i, subnet in enumerate(self.private_subnets):
            aws.ec2.RouteTableAssociation(
                f"payment-private-rta-{i+1}-{args.environment_suffix}",
                subnet_id=subnet.id,
                route_table_id=self.private_route_table.id,
                opts=child_opts
            )

        # Register outputs
        self.register_outputs({
            'vpc_id': self.vpc.id,
            'vpc_cidr': self.vpc.cidr_block,
            'public_subnet_ids': [s.id for s in self.public_subnets],
            'private_subnet_ids': [s.id for s in self.private_subnets],
            'nat_gateway_id': self.nat_gateway.id,
        })
